src/tree/updater_quantile_hist.h

```diff
@@ -28,6 +28,38 @@
 #include "../common/column_matrix.h"
 
 namespace xgboost {
+
+template<typename T, size_t MaxStackSize>
```

Comment:
```suggestion
/*!
 * \brief A C-style array with in-stack allocation. As long as the array is smaller than MaxStackSize, it will be allocated inside the stack. Otherwise, it will be heap-allocated.
 */
template<typename T, size_t MaxStackSize>
```

Reply:
@SmirnovEgorRu Is stack allocation more performant than heap allocation?

Reply:
@hcho3 Yes, stack allocation should be more performant, help allocation requires some searching/merging to find available memory block.  `malloc` itself manages some sort of memory pool.

Reply:
@hcho3, yes, because even with single-thread application stack allocation - really cheap.
But most important point here that in multi-threaded application calling of common system malloc() in different threads at the same moment leads to threads locking. In some workloads it can became really hotspot.

For memory allocation with threading better to use tbb_malloc or jemalloc - they are pretty good in this case. In XGBoost we don't have them, so I used stack allocator for small buffers.

Some more details about advantages of scalable mallocs can be found here:
https://www.infoworld.com/article/3201285/why-effective-parallel-programming-must-include-scalable-memory-allocation.html